File: backend/src/main/java/com/styp/cenate/service/mbac/AuditoriaService.java
package com.styp.cenate.service.mbac;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import com.styp.cenate.dto.mbac.AuditoriaModularResponseDTO;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;

/**
 * Interfaz de servicio para la gestión de auditoría modular.
 * Proporciona métodos para consultar los registros de auditoría de permisos.
 * 
 * @author CENATE Development Team
 * @version 1.0
 */
public interface AuditoriaService {

    /**
     * Obtiene toda la auditoría modular paginada.
     * 
     * @param pageable Información de paginación
     * @return Página de registros de auditoría
     */
    Page<AuditoriaModularResponseDTO> obtenerAuditoriaModular(Pageable pageable);

    /**
     * Obtiene la auditoría de un usuario específico.
     * 
     * @param userId ID del usuario
     * @param pageable Información de paginación
     * @return Página de registros de auditoría del usuario
     */
    Page<AuditoriaModularResponseDTO> obtenerAuditoriaPorUsuario(Long userId, Pageable pageable);

    /**
     * Obtiene la auditoría por nombre de usuario.
     * 
     * @param username Nombre de usuario
     * @param pageable Información de paginación
     * @return Página de registros de auditoría del usuario
     */
    Page<AuditoriaModularResponseDTO> obtenerAuditoriaPorUsername(String username, Pageable pageable);

    /**
     * Obtiene la auditoría de un módulo específico.
     * 
     * @param modulo Nombre del módulo
     * @param pageable Información de paginación
     * @return Página de registros de auditoría del módulo
     */
    Page<AuditoriaModularResponseDTO> obtenerAuditoriaPorModulo(String modulo, Pageable pageable);

    /**
     * Obtiene la auditoría por tipo de acción.
     * 
     * @param accion Acción realizada (INSERT, UPDATE, DELETE)
     * @param pageable Información de paginación
     * @return Página de registros de auditoría de la acción
     */
    Page<AuditoriaModularResponseDTO> obtenerAuditoriaPorAccion(String accion, Pageable pageable);

    /**
     * Obtiene la auditoría en un rango de fechas.
     * 
     * @param fechaInicio Fecha de inicio
     * @param fechaFin Fecha de fin
     * @param pageable Información de paginación
     * @return Página de registros de auditoría en el rango especificado
     */
    Page<AuditoriaModularResponseDTO> obtenerAuditoriaPorRangoFechas(
            LocalDateTime fechaInicio, 
            LocalDateTime fechaFin, 
            Pageable pageable
    );

    /**
     * Obtiene la auditoría de un usuario en un rango de fechas.
     * 
     * @param userId ID del usuario
     * @param fechaInicio Fecha de inicio
     * @param fechaFin Fecha de fin
     * @param pageable Información de paginación
     * @return Página de registros de auditoría
     */
    Page<AuditoriaModularResponseDTO> obtenerAuditoriaPorUsuarioYFechas(
            Long userId, 
            LocalDateTime fechaInicio, 
            LocalDateTime fechaFin, 
            Pageable pageable
    );

    /**
     * Obtiene la auditoría por módulo y acción.
     * 
     * @param modulo Nombre del módulo
     * @param accion Acción realizada
     * @param pageable Información de paginación
     * @return Página de registros de auditoría
     */
    Page<AuditoriaModularResponseDTO> obtenerAuditoriaPorModuloYAccion(
            String modulo, 
            String accion, 
            Pageable pageable
    );

    /**
     * Obtiene un resumen de eventos por tipo.
     * 
     * @return Mapa con el conteo de eventos por tipo
     */
    Map<String, Long> obtenerResumenPorTipoEvento();

    /**
     * Obtiene los últimos N eventos de auditoría.
     * 
     * @param limit Número de eventos a obtener
     * @return Lista de los últimos eventos
     */
    List<AuditoriaModularResponseDTO> obtenerUltimosEventos(int limit);

    /**
     * Busca en la auditoría por texto en el detalle.
     *
     * @param searchText Texto a buscar
     * @param pageable Información de paginación
     * @return Página de registros que coinciden con la búsqueda
     */
    Page<AuditoriaModularResponseDTO> buscarEnDetalle(String searchText, Pageable pageable);

    /**
     * Obtiene sesiones activas (usuarios conectados).
     *
     * @return Lista de mapas con información de usuarios conectados
     */
    List<Map<String, Object>> obtenerSesionesActivas();

    /**
     * Cierra forzadamente una sesión activa de un usuario.
     * Crea un evento de LOGOUT en la auditoría.
     *
     * @param usuarioSesion ID del usuario cuya sesión se cerrará
     * @return true si se cerró exitosamente, false si no había sesión activa
     */
    boolean cerrarSesionUsuario(String usuarioSesion);
}

File: frontend/src/services/auditoriaService.js
// ========================================================================
// 📊 auditoriaService.js – Servicio para Auditoría MBAC (CENATE 2025)
// ------------------------------------------------------------------------
// Servicio para consultar logs y auditoría del sistema
// ========================================================================

import { apiClient } from "../../lib/apiClient";

const auditoriaService = {
  /**
   * Obtener resumen de auditoría
   * @returns {Promise} Resumen de logs
   */
  async getResumen() {
    try {
      const response = await apiClient.get("/auditoria/resumen", true);
      return response;
    } catch (error) {
      console.error("Error al obtener resumen de auditoría:", error);
      throw error;
    }
  },

  /**
   * Obtener últimos logs
   * @param {number} limit - Número de logs a obtener
   * @returns {Promise} Array de logs
   */
  async getUltimos(limit = 50) {
    try {
      const response = await apiClient.get(`/auditoria/ultimos?limit=${limit}`, true);
      return response;
    } catch (error) {
      console.error("Error al obtener últimos logs:", error);
      throw error;
    }
  },

  /**
   * Obtener logs por usuario
   * @param {number} userId - ID del usuario
   * @returns {Promise} Array de logs del usuario
   */
  async getByUsuario(userId) {
    try {
      const response = await apiClient.get(`/auditoria/usuario/${userId}`, true);
      return response;
    } catch (error) {
      console.error("Error al obtener logs del usuario:", error);
      throw error;
    }
  },

  /**
   * Obtener logs por username
   * @param {string} username - Username del usuario
   * @returns {Promise} Array de logs del usuario
   */
  async getByUsername(username) {
    try {
      const response = await apiClient.get(`/auditoria/username/${username}`, true);
      return response;
    } catch (error) {
      console.error("Error al obtener logs por username:", error);
      throw error;
    }
  },

  /**
   * Obtener logs por rango de fechas
   * @param {string} fechaInicio - Fecha inicio (ISO format)
   * @param {string} fechaFin - Fecha fin (ISO format)
   * @returns {Promise} Array de logs en el rango
   */
  async getByRango(fechaInicio, fechaFin) {
    try {
      const response = await apiClient.get(
        `/auditoria/rango?inicio=${fechaInicio}&fin=${fechaFin}`,
        true
      );
      return response;
    } catch (error) {
      console.error("Error al obtener logs por rango:", error);
      throw error;
    }
  },

  /**
   * Obtener logs por módulo
   * @param {string} modulo - Nombre del módulo
   * @returns {Promise} Array de logs del módulo
   */
  async getByModulo(modulo) {
    try {
      const response = await apiClient.get(`/auditoria/modulo/${modulo}`, true);
      return response;
    } catch (error) {
      console.error("Error al obtener logs del módulo:", error);
      throw error;
    }
  },

  /**
   * Obtener logs por acción
   * @param {string} accion - Tipo de acción
   * @returns {Promise} Array de logs de la acción
   */
  async getByAccion(accion) {
    try {
      const response = await apiClient.get(`/auditoria/accion/${accion}`, true);
      return response;
    } catch (error) {
      console.error("Error al obtener logs por acción:", error);
      throw error;
    }
  },

  /**
   * Buscar en logs
   * @param {string} query - Término de búsqueda
   * @returns {Promise} Array de logs que coinciden
   */
  async buscar(query) {
    try {
      const response = await apiClient.get(`/auditoria/buscar?q=${encodeURIComponent(query)}`, true);
      return response;
    } catch (error) {
      console.error("Error al buscar en logs:", error);
      throw error;
    }
  },

  /**
   * Obtener módulos auditados
   * @returns {Promise} Lista de módulos con auditoría
   */
  async getModulos() {
    try {
      const response = await apiClient.get("/auditoria/modulos", true);
      return response;
    } catch (error) {
      console.error("Error al obtener módulos auditados:", error);
      throw error;
    }
  },

  /**
   * Health check del servicio de auditoría
   * @returns {Promise} Estado del servicio
   */
  async healthCheck() {
    try {
      const response = await apiClient.get("/auditoria/health", true);
      return response;
    } catch (error) {
      console.error("Error en health check de auditoría:", error);
      throw error;
    }
  },

  /**
   * Obtener logs recientes (últimos 10)
   * @returns {Promise} Array de logs recientes
   */
  async getLogsRecientes() {
    try {
      const response = await apiClient.get("/admin/audit/logs/recientes", true);
      return Array.isArray(response) ? response : [];
    } catch (error) {
      console.error("Error al obtener logs recientes:", error);
      return [];
    }
  },
};

export default auditoriaService;
